frame/offences/benchmarking/src/lib.rs

```diff
@@ -151,25 +178,162 @@ benchmarks! {
 		};
-
+		assert_eq!(System::<T>::event_count(), 0);
 	}: {
 		let _ = <T as ImOnlineTrait>::ReportUnresponsiveness::report_offence(reporters, offence);
 	}
+	verify {
+		// make sure the report was not deferred
+		assert!(Offences::<T>::deferred_offences().is_empty());
+		// make sure that all slashes have been applied
+		assert_eq!(
+			System::<T>::event_count(), 0
+			+ 1 // offence
+			+ 2 * r // reporter (reward + endowment)
+			+ o // offenders slashed
+			+ o * n // nominators slashed
+		);
+	}
+
+	report_offence_grandpa {
+		let r in 1 .. MAX_REPORTERS;
+		let n in 0 .. MAX_NOMINATORS.min(MAX_NOMINATIONS as u32);
+		let o = 1;
+
+		// Make r reporters
+		let mut reporters = vec![];
+		for i in 0 .. r {
+			let reporter = account("reporter", i, SEED);
+			reporters.push(reporter);
+		}
+
+		// make sure reporters actually get rewarded
+		Staking::<T>::set_slash_reward_fraction(Perbill::one());
+
+		let mut offenders = make_offenders::<T>(o, n).expect("failed to create offenders");
+		let keys =  ImOnline::<T>::keys();
+
+		let offence = GrandpaEquivocationOffence {
+			time_slot: GrandpaTimeSlot { set_id: 0, round: 0 },
+			session_index: 0,
+			validator_set_count: keys.len() as u32,
+			offender: T::convert(offenders.pop().unwrap()),
+		};
+		assert_eq!(System::<T>::event_count(), 0);
+	}: {
+		let _ = Offences::<T>::report_offence(reporters, offence);
+	}
+	verify {
+		// make sure the report was not deferred
+		assert!(Offences::<T>::deferred_offences().is_empty());
+		// make sure that all slashes have been applied
+		assert_eq!(
+			System::<T>::event_count(), 0
+			+ 1 // offence
+			+ 2 * r // reporter (reward + endowment)
+			+ o // offenders slashed
+			+ o * n // nominators slashed
+		);
+	}
+
+	report_offence_babe {
+		let r in 1 .. MAX_REPORTERS;
+		let n in 0 .. MAX_NOMINATORS.min(MAX_NOMINATIONS as u32);
+		let o = 1;
+
+		// Make r reporters
+		let mut reporters = vec![];
+		for i in 0 .. r {
+			let reporter = account("reporter", i, SEED);
+			reporters.push(reporter);
+		}
+
+		// make sure reporters actually get rewarded
+		Staking::<T>::set_slash_reward_fraction(Perbill::one());
+
+		let mut offenders = make_offenders::<T>(o, n).expect("failed to create offenders");
+		let keys =  ImOnline::<T>::keys();
+
+		let offence = BabeEquivocationOffence {
+			slot: 0,
+			session_index: 0,
+			validator_set_count: keys.len() as u32,
+			offender: T::convert(offenders.pop().unwrap()),
+		};
+		assert_eq!(System::<T>::event_count(), 0);
+	}: {
+		let _ = Offences::<T>::report_offence(reporters, offence);
+	}
+	verify {
+		// make sure the report was not deferred
+		assert!(Offences::<T>::deferred_offences().is_empty());
+		// make sure that all slashes have been applied
+		assert_eq!(
+			System::<T>::event_count(), 0
+			+ 1 // offence
+			+ 2 * r // reporter (reward + endowment)
+			+ o // offenders slashed
+			+ o * n // nominators slashed
+		);
+	}
 
 	on_initialize {
-		let d in ...;
+		let d in 1 .. MAX_DEFERRED_OFFENCES;
```

Comment:
I had such question when benchmarking another module: Is it worth even making this a complexity parameter here? because it is part of state and rule of thump that we cannot read state for weight computation. Even if there is a correlation, not much that we can do about it right? why not just put some average constant in there?

Reply:
maybe the final analysis formula has some details that I don't know.

Reply:
I think we can at least observe if the increase is linear or not. In case it's non-linear it will probably make sense to limit the number of deferred offences or process them in batches instead of all at once.